src/layers/dynamic-opacity-scatterplot-layer-shaders.js

```diff
@@ -0,0 +1,152 @@
+  DECKGL_FILTER_SIZE(offset, geometry);
+  gl_Position = project_position_to_clipspace(instancePositions, instancePositions64Low, offset, geometry.position);
+  DECKGL_FILTER_GL_POSITION(gl_Position, geometry);
+
+  // Apply opacity to instance color, or return instance picking color
+  vFillColor = vec4(instanceFillColors.rgb, opacity);
+  DECKGL_FILTER_COLOR(vFillColor, geometry);
+  vLineColor = vec4(instanceLineColors.rgb, opacity);
+  DECKGL_FILTER_COLOR(vLineColor, geometry);
+
+  vExpressionValue = instanceExpressionValue / 255.0;
+  vSelectionState = instanceSelectionState;
+}
+`;
+
+/**
+ * Fragment shader adapted to perform aggregation and
+ * take color scale functions + sliders into account.
+ * Reference: https://github.com/visgl/deck.gl/blob/8.4-release/modules/layers/src/scatterplot-layer/scatterplot-layer-fragment.glsl.js
+ * Reference: https://observablehq.com/@rreusser/selecting-the-right-opacity-for-2d-point-clouds
+ * Reference: https://github.com/flekschas/regl-scatterplot/blob/master/src/point.fs
+ * Reference: https://github.com/hubmapconsortium/vitessce-image-viewer/blob/06231ae02cac1ff57ba458c71e9bc59ed2fc4f8b/src/layers/XRLayer/xr-layer-fragment-colormap.webgl1.glsl
+ */
+export const fragmentShader = glsl`
+#define SHADER_NAME dynamic-opacity-scatterplot-layer-fragment-shader
+
+precision highp float;
+
+#pragma glslify: plasma = require("glsl-colormap/plasma")
+#pragma glslify: viridis = require("glsl-colormap/viridis")
+#pragma glslify: greys = require("glsl-colormap/greys")
+#pragma glslify: magma = require("glsl-colormap/magma")
+#pragma glslify: jet = require("glsl-colormap/jet")
+#pragma glslify: bone = require("glsl-colormap/bone")
+#pragma glslify: copper = require("glsl-colormap/copper")
+#pragma glslify: density = require("glsl-colormap/density")
+#pragma glslify: inferno = require("glsl-colormap/inferno")
+#pragma glslify: cool = require("glsl-colormap/cool")
+#pragma glslify: hot = require("glsl-colormap/hot")
+#pragma glslify: spring = require("glsl-colormap/spring")
+#pragma glslify: summer = require("glsl-colormap/summer")
+#pragma glslify: autumn = require("glsl-colormap/autumn")
+#pragma glslify: winter = require("glsl-colormap/winter")
+
+
+// Custom uniforms for Vitessce:
+uniform float opacity;
+uniform vec2 uColorScaleRange;
+uniform bool uIsExpressionMode;
+
+
+varying vec2 unitPosition;
+varying float innerUnitRadius;
+varying float outerRadiusPixels;
+varying vec4 vFillColor;
+// Custom varyings for Vitessce:
+varying float vExpressionValue;
+varying float vSelectionState;
+
+
+void main(void) {
+  geometry.uv = unitPosition;
+  
+  float distToCenter = length(unitPosition) * outerRadiusPixels;
+  float inCircle = smoothedge(distToCenter, outerRadiusPixels);
+
+  if (inCircle == 0.0) {
+    discard;
+  }
+
+  float scaledExpressionValue = (vExpressionValue - uColorScaleRange[0]) / max(0.005, (uColorScaleRange[1] - uColorScaleRange[0]));
+  if(uIsExpressionMode) {
```

Comment:
Performance could suffer in the future with this because if-then statements can be a bit risky in shaders.  Not sure it's worth changing at the moment, but good to watch out for (and have documented in this PR in the form of this comment).  Usually I do something like

```
gl_FragColor = float(uIsExpressionMode) * __colormap(clamp(scaledExpressionValue, 0.0, 1.0)) +  (1. - float(uIsExpressionMode)) * vFillColor; 
```

Reply:
Thank you!